mssqlcli/main.py

```diff
@@ -594,14 +595,40 @@ def refresh_completions(self, history=None, persist_priorities='all'):
         :param persist_priorities: 'all' or 'keywords'
         """
 
+        mssqlclclient_completion_refresher = self.get_completion_refresher_mssqlcliclient(
+                                                self.mssqlcliclient_query_execution)
+
         callback = functools.partial(self._on_completions_refreshed,
                                      persist_priorities=persist_priorities)
 
-        self.completion_refresher.refresh(self.mssqlcliclient_query_execution,
-                                          callback, history=history, settings=self.settings)
+        self.completion_refresher.refresh(mssqcliclient=mssqlclclient_completion_refresher,
+                                          callbacks=callback,
+                                          history=history,
+                                          settings=self.settings)
+
+
+
         return [(None, None, None,
                  'Auto-completion refresh started in the background.')]
 
+    def get_completion_refresher_mssqlcliclient(self, mssqlcliclient):
+        mssqlcliclient_completion_refreshser = MssqlCliClient(

```

Comment:
>MssqlCliClient [](start = 47, length = 14)

Perhaps we introduce a copy method to MssqlCliClient to centralize copying an instance?

Reply:
I originally thought that would be a good idea, but then OOP wise I don't think a mssqlclient should be responsible for copying itself and generating a new owner URI. I moved that responsibility up a level. Let me know your thoughts.

Reply:
Two options: 

1) we currently generate an owner_id in the constructor, so can generate a new owner_id as part of the copy operation. 

2) we can pass in the owner_id in the copy method.

I think I like #2, and if the value not passed, we create it:

if not owner_uri:
            self.owner_uri = generate_owner_uri()

Maybe clone would be a better name to avoid shallow/deep copy confusion.

---
In reply to: [167103472](https://github.com/dbcli/mssql-cli/pull/161#discussion_r167103472) [](ancestors = 167103472)